### PR TITLE
[API] New jupyter notebook: images dataset creation, visualization and classification

### PR DESCRIPTION
**Pull request by spell00**
New Jupyter notebook to display examples on how to create datasets and basic visualization and classification methods

### CODE REVIEW

**Review by christinerogers — CHANGES_REQUESTED** [review]
Hi @spell00, to start -- Could you describe at top a few more details on what the purpose, input and output are? 
.

**Review by xlecours — COMMENTED** [review]
I can't see the notebook in github. Was it generated with Jupyter Notebook?

**Comment by spell00** [comment]
It was generated with Google Colab. It is working for me though, I see the notebook in Github.

**Comment by spell00** [comment]
blocked by #7101

**Comment by christinerogers** [comment]
@spell00 could you remind me - was there a colab link for this one?

**Comment by spell00** [comment]
Yes. There is actually 2 ways of accessing the colab notebook. The first one is to used the [link](https://colab.research.google.com/drive/1m5-6lCYDy0XSlYd4pFPxVpvzTzz3JW50?authuser=1#scrollTo=9YPEDWfq7Zyw) to the notebook in my google drive (for my mcin account). The second way, which I think is better, is to use the address of the notebook in the Github repository. In this case, this is the [link](https://colab.research.google.com/github/spell00/Loris/blob/2020-08-06-JupyterCreateImageDataset/docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb) . For testing purposes, it links to my branch of Loris, but it would be better to change it to the file in `main` before merging the notebook.

File: docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb
Comment (cmadjar) [inline docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb:47]:
```suggestion
        "# Image analysis with fMRI 3D images imported with LORIS API"
```

File: docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb
Comment (cmadjar) [inline docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb:56]:
```suggestion
        "This is a tutorial to show how to use Loris' API to download MRI images. It also contains a few examples of how the data can be used to run basic data analysis. \n",
```

File: docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb
Comment (cmadjar) [inline docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb:94]:
```suggestion
        "import json           # Provide convenient functions to handle json objects \n",
```

File: docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb
Comment (cmadjar) [inline docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb:227]:
```suggestion
        "We use the URLs we got in the previous step to download all T1 and T2 MRI images. The images are stored in a dictionary in their according groups (T1 or T2) for later.\n",
```

Comment (cmadjar) [reply docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb:227]: note: T1 and T2 are anatomical MRIs, not function MRIs (fMRI)

File: docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb
Comment (cmadjar) [inline docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb:342]:
```suggestion
        "In this tutorial, T1 and T2 images are compared. They are of similar sizes (160x256x224 and 160x256x256 for T1 and T2, respectively), but they need to be exactly the same size for any subsequent analysis.\n",
        "\n",
        "In machine learning, it is common practice to reduce large images before training a model. Large images have the advantage of containing more information, but it comes with a tradeoff known as the Curse of dimensionality. Having a high dimensionality can make it much easier to have good performances on the training set, but the models trained overfit more easily to the training data and perform poorly on the validation and test data.\n",
        "\n",
        "Of course, reducing images too much will also harm the performance of the model trained. There is no rule of thumb or algorithm to get the optimal size of images to be used in a specific task, so it might be a good idea to try a few different reductions.\n",
        "\n",
        "This tutorial will explore 2 dimensions. Both will cubes (all sides have the same length): 128x128x128 and 32x32x32. The later dimensions might be a huge reduction, but the 3D images still have 32,768 dimensions (each voxel being a dimension), which is still huge, but much more manageable than the larger reduction, which has 2,097,152 dimensions. In order to decide which reduction to use, we will observe the data using a Principal Component Analysis (PCA). It will give an idea of whether the data has lost too much information to use it in a classification task. \n",
        "\n",
        "Ultimately, it might be necessary to use both strategies to test if one is better than the other. In case both strategies appear to be equal, Ockham's razor principle suggest the images with fewer voxels should be used. In this case, the notion of equality is somewhat arbitrary and might depend on the task to be accomplished."
```

**Review by cmadjar — CHANGES_REQUESTED** [review]
@spell00 Very nice work!!

I would recommend a few changes before we merge this though:
- add a README.md file in the notebook directory to document how to install and use the Jupyter notebook as well as what is the Jupyter notebook. 
- in the README.md, ensure to put those dependencies and installation instructions (they can be better phrased than what I propose here):

  - `pip` dependencies:
```
pip install notebook
pip install ipywidgets  # needed for the progress bar
```
  - And then run the following to enable the `ipywidgetnbextension` in the notebook (needed for the progress bar): 
```
jupyter nbextension enable --py widgetsnbextension
```
  - Then, type in to run jupyter Notebook in the current working directory:
```
jupyter notebook
```

- missing one dependency in the pip install portion of the Jupyter notebook (h5py, which is needed to read MINC files)

- several suggestions on the documentation within the Jupyter notebook

File: docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb
Comment (spell00) [reply docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb:100]: are you sure installing `ipywidgets` and enabling widgets is necessary? I tested the notebook on a fresh anaconda/python 3.8 installation and the progress bar is fine. Here is how it looks without any modifications to the notebook:

pbar

File: docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb
Comment (spell00) [reply docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb:73]: Why should it be installed, I don't use it and I have no error anywhere. I can import the images, transform them in numpy tensors and display the images in the notebook just fine. 

mri_exemple

**Review by spell00 — COMMENTED** [review]
Thanks @cmadjar for the review & comments!

File: docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb
Comment (spell00) [reply docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb:100]: I think I should also add the option of installing `anaconda` and use `Anaconda Navigator` in the README you are suggesting to add.

File: docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb
Comment (cmadjar) [reply docs/notebooks/LORIS-API_Part3-Create_image_dataset.ipynb:73]: weird. maybe this is a Mac thing. Unless the files you are downloading are NIfTI files and not MINC? I was not fully surprised by the name of the package since MINC is based on the hdf5 format and h5py is HDF5 for python.

**Comment by cmadjar** [comment]
@spell00 @xlecours This is very nice work but I am not sure this belongs to here given that this is a example Jupyter notebook to execute processing pipeline. If the notebook was only about querying the API, then, I agree, it could be added to the LORIS repo but given that this is an analysis pipeline for T1 and T2, I don't think this belongs here.

However, @xlecours mentioned adding this to Neurohub instead when I talked to him about this. Might be worth exploring that option.

Again, this is a very nice work @spell00

**Comment by driusan** [comment]
closed as per @cmadjar's latest comment. Will find new home for applicable notebooks.

**Pull request closed without merging**
